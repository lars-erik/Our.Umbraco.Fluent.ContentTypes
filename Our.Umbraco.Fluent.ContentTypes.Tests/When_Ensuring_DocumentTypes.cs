using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using System.Text;
using System.Threading.Tasks;
using ApprovalTests;
using ApprovalTests.Reporters;
using AutoMapper;
using Moq;
using Newtonsoft.Json;
using NUnit.Framework;
using Umbraco.Core;
using Umbraco.Core.Cache;
using Umbraco.Core.Logging;
using Umbraco.Core.Models;
using Umbraco.Core.Models.Mapping;
using Umbraco.Core.ObjectResolution;
using Umbraco.Core.PropertyEditors;
using Umbraco.Tests.BootManagers;
using Umbraco.Tests.Models.Mapping;
using Umbraco.Web.Models.ContentEditing;

namespace Our.Umbraco.Fluent.ContentTypes.Tests
{
    [TestFixture]
    [UseReporter(typeof(VisualStudioReporter))]
    public class When_Ensuring_DocumentTypes : ComparisonTestBase
    {
        private PropertyEditorResolver editorResolver;

        [Test]
        public void Prevents_Circular_References()
        {
            Config.DocumentType("a")
                .Parent("b");

            Config.DocumentType("b")
                .AllowedChildren("a");

            Assert.That(() =>
                Config.Ensure(Config.Compare()),
                Throws.Exception.InnerException.With
                    .Message.EqualTo("Seems like we've hit a rift in time-space continuum. (Circular reference)")
            );
        }

        [Test]
        public void Creates_New()
        {
            StubDataType(50, "RTE");
            StubDataType(51, "Checkbox");

            StubTemplate("template", new Guid("73B57DF0-F0F4-4F9A-992E-C2A4F8AB8346"));

            StubContentType(1, "parent");
            StubContentType(2, "child");
            StubContentType(3, "mixin");

            Config.DocumentType("docType")
                .Name("A nice doctype")
                .Description("A fine description")
                .Icon("icon-folder")
                .AllowedAsRoot(true)
                .Parent("parent")
                .AllowedTemplates("template")
                .AllowedChildren("child")
                .Compositions("mixin")
                .Tab("A tab")
                    .Property("richText")
                        .DisplayName("Killer article")
                        .Description("Go wild here")
                        .DataType("RTE")
                    .Property("blockRobots")
                        .DisplayName("Block robots")
                        .Description("Chicken out")
                        .DataType("Checkbox")
                .Tab("Secret tab")
                    .Property("makeMoney")
                        .DisplayName("Make money from this")
                        .Description("Tick here to make immense amounts of money")
                        .DataType("Checkbox");

            Config.Ensure(Config.Compare());

            Mock.Get(Support.ServiceContext.ContentTypeService)
                .Verify(s => s.Save(Match.Create<IContentType>(VerifyContentType), 0));
        }

        [Test]
        public void Orders_Dependent_Types_Later_Than_Their_Dependencies()
        {
            /*
            Setup

            Alias    Parent    AllowedChildren    Compositions      Dependencies           Indirect                      All
            d-1-1    d-1;                         cmp-1-1; cmp-2;   d-1; cmp-1-1; cmp-2;   cmp-1;                        d-1; cmp-1-1; cmp-2; cmp-1; 
            cmp-1-1                               cmp-1;            cmp-1;                                               cmp-1; 
            cmp-2                        
            d-2                d-1; d-1-1;                          d-1; d-1-1;            d-1; cmp-1-1; cmp-2; cmp-1;   d-1; d-1-1; d-1; cmp-1-1; cmp-2; cmp-1; 
            cmp-1                                                                                  
            d-1-2    d-1;                                           d-1;                   cmp-1;                        d-1; cmp-1; 
            d-1                                   cmp-1;            cmp-1;                                               cmp-1; 

            Expectation
            cmp-1
            cmp-2
            cmp-1-1
            d-1
            d-1-1
            d-1-2
            d-2

            */


            Config.DocumentType("d-1-1")
                .Parent("d-1")
                .Compositions("cmp-1-1", "cmp-2");

            Config.DocumentType("cmp-1-1")
                .Compositions("cmp-1");

            Config.DocumentType("cmp-2");

            Config.DocumentType("d-2")
                .AllowedAsRoot(true)
                .AllowedChildren("d-1", "d-1-1");

            Config.DocumentType("cmp-1");

            Config.DocumentType("d-1-2")
                .Parent("d-1");

            Config.DocumentType("d-1")
                .Compositions("cmp-1");
            //.AllowedChildren("d-1-1"); // dammit! - what about circular references? (parent/allowedchildren) - rare UC?
            // See prevents circular reference test.

            var diff = Config.Compare();

            var orderedTypes = DependencyComparer.OrderByDependencies(diff.DocumentTypes.Values);
            var aliases = orderedTypes.Select(t => t.Key);

            Approvals.VerifyJson(aliases.ToJson(Formatting.None));
        }

        [SetUp]
        public void SetupMapping()
        {
            Func<IEnumerable<Type>> typeProducer = Enumerable.Empty<Type>;
            var _propertyEditorResolver = new Mock<PropertyEditorResolver>(
                Mock.Of<IServiceProvider>(), 
                Mock.Of<ILogger>(), 
                typeProducer, 
                (IRuntimeCacheProvider)CacheHelper.CreateDisabledCacheHelper().RuntimeCache
            );
            editorResolver = _propertyEditorResolver.Object;

            var ctor = Type.GetType("Umbraco.Web.Models.Mapping.ContentTypeModelMapper, umbraco")
                .GetConstructor(new[] { typeof(Lazy<PropertyEditorResolver>) });
            var entityctor = Type.GetType("Umbraco.Web.Models.Mapping.EntityModelMapper, umbraco")
                .GetConstructor(new Type[0]);
            Mapper.Initialize(configuration =>
            {
                var mapper = (MapperConfiguration)ctor.Invoke(new[] { new Lazy<PropertyEditorResolver>(() => editorResolver) });
                mapper.ConfigureMappings(configuration, Support.UmbracoContext.Application);

                var entityMapper = (MapperConfiguration)entityctor.Invoke(new object[0]);
                entityMapper.ConfigureMappings(configuration, Support.UmbracoContext.Application);
            });

            Mock.Get(editorResolver).Setup(r => r.GetByAlias(It.IsAny<string>())).Returns(new TestablePropertyEditor());
            Mock.Get(Support.ServiceContext.DataTypeService)
                .Setup(s => s.GetPreValuesCollectionByDataTypeId(It.IsAny<int>()))
                .Returns(new PreValueCollection(new Dictionary<string, PreValue>()));
        }

        private bool VerifyContentType(IContentType obj)
        {
            var jsonable = Mapper.Map<DocumentTypeDisplay>(obj);
            jsonable.Key = new Guid("2BFEC211-884B-4E70-A8BD-6992C0DC687C");
            jsonable.Udi = new GuidUdi("document-type", jsonable.Key);
            Approvals.VerifyJson(jsonable.ToJson());
            return true;
        }

        [PropertyEditor("dummy", "Dummy", "is-nice.html")]
        public class TestablePropertyEditor : PropertyEditor
        {
        }
    }
}
