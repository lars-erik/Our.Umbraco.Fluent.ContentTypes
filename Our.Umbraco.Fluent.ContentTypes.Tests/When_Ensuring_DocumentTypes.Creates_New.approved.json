{
    "allowedTemplates": [
        {
            "name": null,
            "id": 0,
            "udi": "umb://template/73b57df0f0f44f9a992ec2a4f8ab8346",
            "icon": "icon-layout",
            "trashed": false,
            "key": "73b57df0-f0f4-4f9a-992e-c2a4f8ab8346",
            "parentId": -1,
            "alias": "template",
            "path": null,
            "metaData": {}
        }
    ],
    "defaultTemplate": {
        "name": null,
        "id": 0,
        "udi": "umb://template/73b57df0f0f44f9a992ec2a4f8ab8346",
        "icon": "icon-layout",
        "trashed": false,
        "key": "73b57df0-f0f4-4f9a-992e-c2a4f8ab8346",
        "parentId": -1,
        "alias": "template",
        "path": null,
        "metaData": {}
    },
    "groups": [
        {
            "contentTypeId": 0,
            "parentTabContentTypes": [],
            "parentTabContentTypeNames": [],
            "properties": [
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "richText",
                    "description": "Go wild here",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Killer article",
                    "sortOrder": 0,
                    "dataTypeId": 50,
                    "groupId": 0
                },
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "blockRobots",
                    "description": "Chicken out",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Block robots",
                    "sortOrder": 0,
                    "dataTypeId": 51,
                    "groupId": 0
                },
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "url",
                    "description": null,
                    "validation": {
                        "mandatory": true,
                        "pattern": "https?://[a-zA-Z0-9-.]+.[a-zA-Z]{2,}"
                    },
                    "label": "Url",
                    "sortOrder": 0,
                    "dataTypeId": 52,
                    "groupId": 0
                }
            ],
            "inherited": false,
            "id": 0,
            "sortOrder": 0,
            "name": "A tab"
        },
        {
            "contentTypeId": 0,
            "parentTabContentTypes": [],
            "parentTabContentTypeNames": [],
            "properties": [
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "makeMoney",
                    "description": "Tick here to make immense amounts of money",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Make money from this",
                    "sortOrder": 0,
                    "dataTypeId": 51,
                    "groupId": 0
                }
            ],
            "inherited": false,
            "id": 0,
            "sortOrder": 1,
            "name": "Secret tab"
        },
        {
            "contentTypeId": 0,
            "parentTabContentTypes": [],
            "parentTabContentTypeNames": [],
            "properties": [
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "richText",
                    "description": "Go wild here",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Killer article",
                    "sortOrder": 0,
                    "dataTypeId": 50,
                    "groupId": -666
                },
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "blockRobots",
                    "description": "Chicken out",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Block robots",
                    "sortOrder": 0,
                    "dataTypeId": 51,
                    "groupId": -666
                },
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "url",
                    "description": null,
                    "validation": {
                        "mandatory": true,
                        "pattern": "https?://[a-zA-Z0-9-.]+.[a-zA-Z]{2,}"
                    },
                    "label": "Url",
                    "sortOrder": 0,
                    "dataTypeId": 52,
                    "groupId": -666
                },
                {
                    "editor": null,
                    "view": "is-nice.html",
                    "config": {},
                    "locked": false,
                    "contentTypeId": 0,
                    "contentTypeName": "A nice doctype",
                    "inherited": false,
                    "id": 0,
                    "alias": "makeMoney",
                    "description": "Tick here to make immense amounts of money",
                    "validation": {
                        "mandatory": false,
                        "pattern": null
                    },
                    "label": "Make money from this",
                    "sortOrder": 0,
                    "dataTypeId": 51,
                    "groupId": -666
                }
            ],
            "inherited": false,
            "id": -666,
            "sortOrder": 999,
            "name": "Generic properties"
        }
    ],
    "isContainer": false,
    "listViewEditorName": "List View - Content",
    "allowedContentTypes": [
        2
    ],
    "compositeContentTypes": [
        "parent",
        "mixin"
    ],
    "lockedCompositeContentTypes": [],
    "allowAsRoot": true,
    "notifications": [],
    "ModelState": null,
    "alias": "docType",
    "updateDate": "0001-01-01T00:00:00",
    "createDate": "0001-01-01T00:00:00",
    "description": "A fine description",
    "thumbnail": "folder.png",
    "iconIsClass": true,
    "iconFilePath": "",
    "thumbnailIsClass": false,
    "thumbnailFilePath": "/umbraco/images/thumbnails/folder.png",
    "name": "A nice doctype",
    "id": 0,
    "udi": "umb://document-type/2bfec211884b4e70a8bd6992c0dc687c",
    "icon": "icon-folder",
    "trashed": false,
    "key": "2bfec211-884b-4e70-a8bd-6992c0dc687c",
    "parentId": 1,
    "path": null,
    "metaData": {}
}